### PR TITLE
[DAG] Teach SelectionDAGBuilder to read parameter alignment of compressstore/expandload.

### PR DESCRIPTION
**Pull request by yetingk**
Previously SelectionDAGBuilder used ABI alignment for compressstore/expandload. This patch allows SelectionDAGBuilder to use parameter alignment like vp intrinsics. This does not follow the original code to default use vector type alignment, since it is possible implemented to unaligned vector alignment.

### CODE REVIEW

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp [hunk 1]
```diff
@@ -4746,6 +4748,8 @@ void SelectionDAGBuilder::visitMaskedLoad(const CallInst &I, bool IsExpanding) {
     // @llvm.masked.expandload.*(Ptr, Mask, Src0)
     Ptr = I.getArgOperand(0);
     Alignment = std::nullopt;
+    if (MaybeAlign Align = I.getParamAlign(0))
+      Alignment = Align;
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:4752]: can just assign Alignment from I.getParamAlign?

Comment (topperc) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:4752]: I think we should be using `I.getParamAlign(0).getValueOr(DAG.getEVTAlign(1))`. The call to `DAG.getEVTAlign(VT)` later doesn't make sense for a compressed store or expanding load. The vector type isn't really relevant. Maybe we would use the alignment for the element type, but 1 is conservatively correct.

Comment (topperc) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp]: To clarify the reason I think the vector type isn't relevant, is that we are like not accessing a full vector of memory. If this intrinsic is used in a loop, the pointer for the next iteration will likely have advanced less than a full vector making it not aligned.

Comment (arsenm) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp]: MaybeAlign shouldn't exist. We should just make all of the contexts behave like getValueOr1

File: llvm/test/CodeGen/X86/masked_expandload_isel.ll [hunk 2]
```diff
@@ -0,0 +1,33 @@
+; RUN: llc < %s -mtriple=x86_64-unknown-linux-gnu -start-after=codegenprepare -stop-before finalize-isel | FileCheck %s
```
Comment (arsenm) [inline llvm/test/CodeGen/X86/masked_expandload_isel.ll:1]: why do you need -start-after?

Comment (arsenm) [reply llvm/test/CodeGen/X86/masked_expandload_isel.ll:1]: Wouldn't that not happen if it's legal for the target?

**Comment by efriedma-quic** [comment]
It's probably worth documenting the behavior in LangRef, like we do for other intrinsics that support "align" markings.

File: llvm/test/CodeGen/X86/masked_compressstore_isel.ll [hunk 3]
```diff
@@ -7,7 +7,23 @@ entry:
...
 }
 
-; CHECK-LABEL: bb.0.entry:
+; CHECK-LABEL:   name: _Z3fooiPiPs
+; CHECK:         %1:vr128x = COPY $xmm1
+; CHECK-NEXT:    %0:vr256x = COPY $ymm0
+; CHECK-NEXT:    %2:vr128x = VPSLLWZ128ri %1, 15
+; CHECK-NEXT:    %3:vk16wm = VPMOVW2MZ128rr killed %2
+; CHECK-NEXT:    %4:vr128x = VPMOVDWZ256rr %0
+; CHECK-NEXT:    VPCOMPRESSWZ128mrk $noreg, 1, $noreg, 0, $noreg, killed %3, killed %4 :: (store unknown-size into `ptr null`, align 1)
```
Comment (topperc) [inline llvm/test/CodeGen/X86/masked_compressstore_isel.ll:16]: Should this have align 16 instead of align 1?

Comment (topperc) [reply llvm/test/CodeGen/X86/masked_compressstore_isel.ll:16]: oh nevermind the check lines are just in a funny spot.

File: llvm/test/CodeGen/X86/masked_expandload_isel.ll (hunk 2 above)
Comment (topperc) [reply llvm/test/CodeGen/X86/masked_expandload_isel.ll:1]: Use update_mir_test_checks.py. I'll fix the compress store test to use the script and we can rebase this on top of that

Comment (topperc) [reply llvm/test/CodeGen/X86/masked_expandload_isel.ll:1]: Please rebase after bdfebc310e67c8e96909e155669277ab75d784ba

**Review by topperc — APPROVED** [review]
LGTM

**Comment by topperc** [comment]
what happened to the LangRef change?

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp [hunk 4]
```diff
@@ -4721,24 +4721,24 @@ void SelectionDAGBuilder::visitMaskedStore(const CallInst &I,
   SDLoc sdl = getCurSDLoc();
 
   auto getMaskedStoreOps = [&](Value *&Ptr, Value *&Mask, Value *&Src0,
-                               MaybeAlign &Alignment) {
+                               Align &Alignment) {
     // llvm.masked.store.*(Src0, Ptr, alignment, Mask)
     Src0 = I.getArgOperand(0);
     Ptr = I.getArgOperand(1);
-    Alignment = cast<ConstantInt>(I.getArgOperand(2))->getMaybeAlignValue();
+    Alignment = cast<ConstantInt>(I.getArgOperand(2))->getAlignValue();
```
Comment (topperc) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:4728]: were any tests passing 0?

**Comment by topperc** [comment]
> > what happened to the LangRef change?
> 
> Sorry, do you mean why I changed LangRef? I changed it because @efriedma-quic asked for. And I opened another PR for this just I think changing ir document is a little exceed the scope of this patch.
> 
> > It's probably worth documenting the behavior in LangRef, like we do for other intrinsics that support "align" markings.

No. I'm asking why Files Changed does not show any change to the LangRef now. I think maybe you lost a commit?

**Review by arsenm — APPROVED** [review]

**Pull request merged**